### DISCUSSION TITLE
Access Client IP from Micro

### DISCUSSION THREAD

**Discussion by lemonyte** [deta/cloud-docs#282] [Help] [answered]
In FastAPI I want to get the client's IP address to implement rate limiting/anti-spam protection. Using `request.client.host` returns `None`, and the `X-Forwarded-For` and `X-Real-IP` headers have the value `99.82.171.163` which looks like the IP address of the Amazon data center where the Micro is hosted. Is there any other way to get the client's IP or is this a technical limitation?
Code:
```python
@app.get('/api/test')
async def test(request: Request):
    return {'ip': request.client.host, 'headers': request.headers}
```
Result:
https://obsidian.deta.dev/api/test
Stack Overflow question:
https://stackoverflow.com/questions/60098005/fastapi-starlette-get-client-real-ip

**Comment by abdelhai** [comment, marked as answer]
unfortunately this is a limitation of amazon global accelerator, which we're using for geo routing. GA doesn't forward the client's IP. we're working out a plan to replace it, nothing concrete yet.

**Reply by mazzz3r** [reply]
Any updates?

**Comment by n0rs3** [comment]
Edit: A follow-up, I resolved my use case using fly.io and deploying a Docker container there, their free tier has shown to be enough so far with both IPv4/6 available 

---

Hello, just adding my comment here so I get a notification when there is an update :slightly_smiling_face: 

Some context regarding why I'm waiting for this implementation and why I need it:

I've currently implemented an API endpoint with Express.js for the [Nextcloudpi][ncp] project using Deta, it checks if the ports 80 & 443 are open and properly configured with port-forwarding and sends a JSON response with data about which ones are open.

I decided to implement a solution for this myself using Deta instead of using external sites as we currently do right now, since they aren't always very happy about having repeated requests on their free services.

For security reason, or mainly just to reduce the potential for abuse, it would be great to be able to fetch the client's IP from the header instead of using it within a POST payload, that way it would only check the ports of the client calling the endpoint which in this case would be the device deploying a [Nextcloudpi][ncp] instance.

[ncp]: https://github.com/nextcloud/nextcloudpi

**Comment by lihugang** [comment]
My solution is following.
1. Buy a domain and set nameservers to Cloudflare. For example, `abc.com`
2. Open your micro and set custom domain to `abc.com`, you will get Deta Server's ip address.
3. Set a `A Record` to the given ip, content is `@`(Root), enable cloudflare's proxy.
4. Now you can get client's ip in `cf-connecting-ip` header.

@lemonyte 

Notice: You must set SSL/TLS encryption mode to `full` in your cloudflare console, because micro enforces https, if it is `flexible`, the connection between cf and micro will use http, and micro will redirect you to https, that's a infinite loop.
